### PR TITLE
Serialization: implement constraint lookups

### PR DESCRIPTION
**Pull request by dannywillems**
Cherry-pick from https://github.com/o1-labs/proof-systems/pull/1966/files.

Used later. Tests coming.

### CODE REVIEW

File: msm/src/serialization/constraints.rs [hunk 1]
```diff
@@ -185,3 +185,71 @@ impl<F: PrimeField> InterpreterEnv<F> for Env<F> {
         }))
     }
 }
+
+impl<Fp: PrimeField> Env<Fp> {
+    #[allow(dead_code)]
+    // FIXME: not mut
+    fn constrain_lookups(&mut self) -> Vec<E<Fp>> {
+        assert_eq!(self.rangecheck4_lookups.len(), 20);
```
Comment (querolita) [inline msm/src/serialization/constraints.rs:193]: This means each row adds always 20 rangecheck4 lookups, and 17 rangecheck15's? Probably this is for the limbs checks?

File: msm/src/serialization/constraints.rs [hunk 2]
```diff
@@ -185,3 +185,71 @@ impl<F: PrimeField> InterpreterEnv<F> for Env<F> {
...
+        assert_eq!(self.rangecheck15_lookups.len(), 17);
+
+        {
+            let rc4_t_lookup = Lookup {
+                table_id: LookupTable::RangeCheck4,
+                numerator: curr_cell(Column::LookupMultiplicity(
```
Comment (querolita) [inline msm/src/serialization/constraints.rs:199]: I am confused...

If I had understood correctly, you need a multiplicity for each table entry, not table id. No?

File: msm/src/serialization/constraints.rs [hunk 3]
```diff
@@ -185,3 +185,71 @@ impl<F: PrimeField> InterpreterEnv<F> for Env<F> {
...
+        // array of lookups
+        // Computing individual "boat"
+        let mut constraints = vec![];
+        let mut idx = 0;
+        for chunk in self.rangecheck4_lookups.chunks(6) {
+            constraints.push(combine_lookups(
+                Column::LookupPartialSum(idx),
+                chunk.to_vec(),
+            ));
+            idx += 1;
```
Comment (querolita) [inline msm/src/serialization/constraints.rs:232]: perhaps add an assert of what's the expected index at this point

File: msm/src/serialization/constraints.rs (hunk 2 above)
Comment (dannywillems) [reply msm/src/serialization/constraints.rs:199]: It is for the expression `m(X)/(\beta + t(X)`

**Review by querolita — APPROVED** [review]

File: msm/src/serialization/constraints.rs (hunk 3 above)
Comment (dannywillems) [reply msm/src/serialization/constraints.rs:232]: Opening https://github.com/o1-labs/proof-systems/issues/1987

File: msm/src/serialization/constraints.rs (hunk 1 above)
Comment (dannywillems) [reply msm/src/serialization/constraints.rs:193]: Yes

File: msm/src/serialization/constraints.rs (hunk 2 above)
Comment (dannywillems) [reply msm/src/serialization/constraints.rs:199]: Explained offline

**Pull request merged**